File: src/coevo/task_flow/mapping.py
"""US-1 stage mapping (AC-7).

Per-unit flow nodes carry an arbitrary ``stage_hint`` string. The
system has a closed set of standardized stages (:class:`StandardStage`)
that all UI and orchestration code speaks. This module turns a
per-unit hint into a :class:`StandardStage` using a deterministic
rule table.

The rule table is itself versioned (the default table is version 1)
so we can later add new standard stages without invalidating
already-confirmed flows.
"""
#
# 中文注释（仅注释，不改逻辑）
# ---------------------------
# US-1 阶段映射：规则表把 unit 节点映射到 StandardStage，O(1) 查询。
from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable

from .models import (
    MappingRule,
    Node,
    ProcessFlow,
    ProcessFlowError,
    StandardStage,
)


@dataclass(frozen=True)
class MappedNode:
    """A per-unit node with its standardized stage attached."""

    node: Node
    standard_stage: StandardStage
    rule_id: str


@dataclass(frozen=True)
class MappedFlow:
    """The mapped view of a :class:`ProcessFlow` (used by downstream)."""

    unit_id: str
    version: int
    nodes: tuple[MappedNode, ...]


def apply_mapping(
    flow: ProcessFlow,
    rules: Iterable[MappingRule] | None = None,
) -> MappedFlow:
    """Map every node in ``flow`` to a :class:`StandardStage`.

    Algorithm
    ---------
    * Sort rules by ``(priority ASC, rule_id ASC)`` and build a
      ``hint -> best rule`` dictionary once (the first rule in the
      sorted order per hint is the winner by definition).
    * For each node, resolve its ``unit_stage_hint`` with a single
      O(1) dictionary lookup instead of scanning the whole table.
    * If no rule matches, raise :class:`ProcessFlowError`.

    The rule table MUST contain :data:`DEFAULT_MAPPING_RULES` plus
    any rules the unit has supplied for itself.
    """
    rule_list = list(rules) if rules is not None else list(DEFAULT_MAPPING_RULES)
    if not rule_list:
        raise ProcessFlowError("mapping rule table must be non-empty")
    sorted_rules = sorted(rule_list, key=lambda r: (r.priority, r.rule_id))
    best_rule_by_hint: dict[str, MappingRule] = {}
    for rule in sorted_rules:
        # The first rule encountered for a hint is the winner because
        # the list is ordered by (priority ASC, rule_id ASC).
        best_rule_by_hint.setdefault(rule.unit_stage_hint, rule)

    mapped: list[MappedNode] = []
    seen: set[str] = set()
    for stage in flow.stages:
        for node in stage.nodes:
            hint = node.stage_hint.value
            match_rule = best_rule_by_hint.get(hint)
            if match_rule is None:
                raise ProcessFlowError(
                    f"no mapping rule matches stage_hint {hint!r} for "
                    f"node {node.node_id!r} in unit {flow.unit_id!r}"
                )
            key = f"{flow.unit_id}:{node.node_id}"
            if key in seen:
                raise ProcessFlowError(f"duplicate node key {key!r} during mapping")
            seen.add(key)
            mapped.append(
                MappedNode(node=node, standard_stage=match_rule.standard_stage,
                           rule_id=match_rule.rule_id)
            )

    if flow.mapping_rules_version != 1:
        raise ProcessFlowError(
            f"unsupported mapping_rules_version {flow.mapping_rules_version!r}; "
            f"only 1 is supported"
        )

    return MappedFlow(
        unit_id=flow.unit_id,
        version=flow.version,
        nodes=tuple(mapped),
    )


#: The default rule table. Versioned to 1.
#:
#: Lower ``priority`` wins ties; ``rule_id`` is the secondary key.
DEFAULT_MAPPING_RULES: tuple[MappingRule, ...] = (
    MappingRule(rule_id="m1",  unit_stage_hint="intake",   standard_stage=StandardStage.INTAKE,   priority=10),
    MappingRule(rule_id="m2",  unit_stage_hint="Intake",   standard_stage=StandardStage.INTAKE,   priority=11),
    MappingRule(rule_id="m3",  unit_stage_hint="接收",      standard_stage=StandardStage.INTAKE,   priority=12),
    MappingRule(rule_id="m4",  unit_stage_hint="接收阶段",  standard_stage=StandardStage.INTAKE,   priority=13),
    MappingRule(rule_id="m10", unit_stage_hint="planning",  standard_stage=StandardStage.PLANNING, priority=10),
    MappingRule(rule_id="m11", unit_stage_hint="Planning",  standard_stage=StandardStage.PLANNING, priority=11),
    MappingRule(rule_id="m12", unit_stage_hint="策划",      standard_stage=StandardStage.PLANNING, priority=12),
    MappingRule(rule_id="m13", unit_stage_hint="方案策划",  standard_stage=StandardStage.PLANNING, priority=13),
    MappingRule(rule_id="m20", unit_stage_hint="execution", standard_stage=StandardStage.EXECUTION, priority=10),
    MappingRule(rule_id="m21", unit_stage_hint="Execution", standard_stage=StandardStage.EXECUTION, priority=11),
    MappingRule(rule_id="m22", unit_stage_hint="执行",      standard_stage=StandardStage.EXECUTION, priority=12),
    MappingRule(rule_id="m23", unit_stage_hint="实施",      standard_stage=StandardStage.EXECUTION, priority=13),
    MappingRule(rule_id="m24", unit_stage_hint="开发阶段",  standard_stage=StandardStage.EXECUTION, priority=14),
    MappingRule(rule_id="m30", unit_stage_hint="review",    standard_stage=StandardStage.REVIEW,   priority=10),
    MappingRule(rule_id="m31", unit_stage_hint="Review",    standard_stage=StandardStage.REVIEW,   priority=11),
    MappingRule(rule_id="m32", unit_stage_hint="审核",      standard_stage=StandardStage.REVIEW,   priority=12),
    MappingRule(rule_id="m33", unit_stage_hint="评审",      standard_stage=StandardStage.REVIEW,   priority=13),
    MappingRule(rule_id="m40", unit_stage_hint="delivery",  standard_stage=StandardStage.DELIVERY, priority=10),
    MappingRule(rule_id="m41", unit_stage_hint="Delivery",  standard_stage=StandardStage.DELIVERY, priority=11),
    MappingRule(rule_id="m42", unit_stage_hint="交付",      standard_stage=StandardStage.DELIVERY, priority=12),
    MappingRule(rule_id="m43", unit_stage_hint="验收",      standard_stage=StandardStage.DELIVERY, priority=13),
    MappingRule(rule_id="m44", unit_stage_hint="提交",      standard_stage=StandardStage.DELIVERY, priority=14),
    MappingRule(rule_id="m50", unit_stage_hint="closure",   standard_stage=StandardStage.CLOSURE,  priority=10),
    MappingRule(rule_id="m51", unit_stage_hint="Closure",   standard_stage=StandardStage.CLOSURE,  priority=11),
    MappingRule(rule_id="m52", unit_stage_hint="收尾",      standard_stage=StandardStage.CLOSURE,  priority=12),
    MappingRule(rule_id="m53", unit_stage_hint="结束",      standard_stage=StandardStage.CLOSURE,  priority=13),
    MappingRule(rule_id="m54", unit_stage_hint="复盘",      standard_stage=StandardStage.CLOSURE,  priority=14),
)

File: go/taskflow/mapping.go
// US-1 stage mapping (AC-7; Go port of src/coevo/task_flow/mapping.py).
//
// Per-unit flow nodes carry an arbitrary StageHint string. The system has a
// closed set of standardized stages (StandardStage) that all UI and
// orchestration code speaks. This file turns a per-unit hint into a
// StandardStage using a deterministic, versioned rule table.
package taskflow

import (
	"fmt"
	"sort"
)

// MappedNode is a per-unit node with its standardized stage attached.
type MappedNode struct {
	Node          Node
	StandardStage StandardStage
	RuleID        string
}

// MappedFlow is the mapped view of a ProcessFlow used by downstream slices.
type MappedFlow struct {
	UnitID  string
	Version int
	Nodes   []MappedNode
}

// DefaultMappingRules returns a defensive copy of the version-1 default rule
// table (27 rules, identical to Python DEFAULT_MAPPING_RULES).
func DefaultMappingRules() []MappingRule {
	return append([]MappingRule(nil), defaultMappingRules...)
}

// defaultMappingRules mirrors Python mapping.DEFAULT_MAPPING_RULES exactly:
// lower Priority wins ties; RuleID is the secondary key.
var defaultMappingRules = []MappingRule{
	{RuleID: "m1", UnitStageHint: "intake", StandardStage: StandardStageIntake, Priority: 10},
	{RuleID: "m2", UnitStageHint: "Intake", StandardStage: StandardStageIntake, Priority: 11},
	{RuleID: "m3", UnitStageHint: "接收", StandardStage: StandardStageIntake, Priority: 12},
	{RuleID: "m4", UnitStageHint: "接收阶段", StandardStage: StandardStageIntake, Priority: 13},
	{RuleID: "m10", UnitStageHint: "planning", StandardStage: StandardStagePlanning, Priority: 10},
	{RuleID: "m11", UnitStageHint: "Planning", StandardStage: StandardStagePlanning, Priority: 11},
	{RuleID: "m12", UnitStageHint: "策划", StandardStage: StandardStagePlanning, Priority: 12},
	{RuleID: "m13", UnitStageHint: "方案策划", StandardStage: StandardStagePlanning, Priority: 13},
	{RuleID: "m20", UnitStageHint: "execution", StandardStage: StandardStageExecution, Priority: 10},
	{RuleID: "m21", UnitStageHint: "Execution", StandardStage: StandardStageExecution, Priority: 11},
	{RuleID: "m22", UnitStageHint: "执行", StandardStage: StandardStageExecution, Priority: 12},
	{RuleID: "m23", UnitStageHint: "实施", StandardStage: StandardStageExecution, Priority: 13},
	{RuleID: "m24", UnitStageHint: "开发阶段", StandardStage: StandardStageExecution, Priority: 14},
	{RuleID: "m30", UnitStageHint: "review", StandardStage: StandardStageReview, Priority: 10},
	{RuleID: "m31", UnitStageHint: "Review", StandardStage: StandardStageReview, Priority: 11},
	{RuleID: "m32", UnitStageHint: "审核", StandardStage: StandardStageReview, Priority: 12},
	{RuleID: "m33", UnitStageHint: "评审", StandardStage: StandardStageReview, Priority: 13},
	{RuleID: "m40", UnitStageHint: "delivery", StandardStage: StandardStageDelivery, Priority: 10},
	{RuleID: "m41", UnitStageHint: "Delivery", StandardStage: StandardStageDelivery, Priority: 11},
	{RuleID: "m42", UnitStageHint: "交付", StandardStage: StandardStageDelivery, Priority: 12},
	{RuleID: "m43", UnitStageHint: "验收", StandardStage: StandardStageDelivery, Priority: 13},
	{RuleID: "m44", UnitStageHint: "提交", StandardStage: StandardStageDelivery, Priority: 14},
	{RuleID: "m50", UnitStageHint: "closure", StandardStage: StandardStageClosure, Priority: 10},
	{RuleID: "m51", UnitStageHint: "Closure", StandardStage: StandardStageClosure, Priority: 11},
	{RuleID: "m52", UnitStageHint: "收尾", StandardStage: StandardStageClosure, Priority: 12},
	{RuleID: "m53", UnitStageHint: "结束", StandardStage: StandardStageClosure, Priority: 13},
	{RuleID: "m54", UnitStageHint: "复盘", StandardStage: StandardStageClosure, Priority: 14},
}

// ApplyMapping maps every node in flow to a StandardStage using rules, or the
// default rule table when rules is nil. Behavior mirrors the Python
// apply_mapping:
//
//   * rules are sorted by (Priority ASC, RuleID ASC); the first rule per hint
//     wins by construction;
//   * each node resolves its hint with a single O(1) lookup;
//   * unknown hints and duplicate unit:node keys fail closed;
//   * only mapping_rules_version 1 is supported.
func ApplyMapping(flow ProcessFlow, rules []MappingRule) (MappedFlow, error) {
	if rules == nil {
		rules = DefaultMappingRules()
	}
	if len(rules) == 0 {
		return MappedFlow{}, Errorf("mapping rule table must be non-empty")
	}
	sorted := append([]MappingRule(nil), rules...)
	sort.Slice(sorted, func(i, j int) bool {
		if sorted[i].Priority != sorted[j].Priority {
			return sorted[i].Priority < sorted[j].Priority
		}
		return sorted[i].RuleID < sorted[j].RuleID
	})
	bestByHint := make(map[string]MappingRule, len(sorted))
	for _, rule := range sorted {
		if _, ok := bestByHint[rule.UnitStageHint]; !ok {
			bestByHint[rule.UnitStageHint] = rule
		}
	}

	seen := make(map[string]struct{})
	var nodes []MappedNode
	for _, stage := range flow.Stages {
		for _, node := range stage.Nodes {
			hintValue, ok := node.StageHint.Value.(string)
			if !ok {
				return MappedFlow{}, Errorf("stage_hint must be a string for node %q in unit %q", node.NodeID, flow.UnitID)
			}
			rule, ok := bestByHint[hintValue]
			if !ok {
				return MappedFlow{}, Errorf("no mapping rule matches stage_hint %q for node %q in unit %q", hintValue, node.NodeID, flow.UnitID)
			}
			key := fmt.Sprintf("%s:%s", flow.UnitID, node.NodeID)
			if _, duplicate := seen[key]; duplicate {
				return MappedFlow{}, Errorf("duplicate node key %q during mapping", key)
			}
			seen[key] = struct{}{}
			nodes = append(nodes, MappedNode{Node: node, StandardStage: rule.StandardStage, RuleID: rule.RuleID})
		}
	}

	if flow.MappingRulesVersion != 1 {
		return MappedFlow{}, Errorf("unsupported mapping_rules_version %d; only 1 is supported", flow.MappingRulesVersion)
	}
	return MappedFlow{UnitID: flow.UnitID, Version: flow.Version, Nodes: nodes}, nil
}
